### PR TITLE
Add details on gradient methods to the quickstart

### PR DESCRIPTION
**Pull request by josh146**
**Context:**

* The quickstarts have been missing information about changing the QNode gradient method, with the only mention of the options being in the QNode docstring. This is for historical reasons, and is essentially holdover from when PennyLane only provided the parameter-shift rule and finite-differences.

* In addition, nowhere in our documentation does it mention that PennyLane works seamlessly with gradient-free optimization methods such as scipy optimize. This is something we get a lot of questions about.

**Description of the Change:**

* Renames the "Interfaces and Training" quickstart to "Gradients and Training"

* Adds a small quickstart on the different gradient methods. This simply lists the allowed methods, with some basic details and where/when they can be used. Note that the goal here is to act like a reference page, so I tried to avoid information overload where possible.

* Added a small example to the NumPy interface showing gradient-free optimization with SciPy.

**Benefits:** see above

**Possible Drawbacks:**

* I noticed that the NumPy interface does not yet mention `requires_grad`. I'm happy to leave this off for a later PR.

**Related GitHub Issues:**

### CODE REVIEW

File: doc/introduction/interfaces.rst [hunk 1]
```diff
@@ -27,6 +109,13 @@ In PennyLane, an interface is declared when creating a :class:`~.QNode`, e.g.,
     If no interface is specified, PennyLane will default to the NumPy interface (powered by the
     `autograd <https://github.com/HIPS/autograd>`_ library).
 
+.. note::
+
+    QNodes with an interface will always incur a small overhead on evaluation. If you do not
+    need to compute quantum gradients of a QNode, specifying ``interface=None`` will remove
+    this overhead and result in a slightly faster evaluation. However, gradients will no
+    longer be available.
+
```
Comment (josh146) [inline doc/introduction/interfaces.rst:118]: I'm on the fence over whether to include this or not. It might be overcomplicated for the quickstart landing page.

File: doc/introduction/interfaces/numpy.rst [hunk 2]
```diff
@@ -326,3 +326,54 @@ of supported and unsupported features. We highlight a few of the major 'gotchas'
...
+
+    params = np.array([0.011, 0.012, 0.05])
+
+    minimize(cost, params, method='BFGS')
+
+Some of the SciPy minimization methods require information about the gradient
+of the cost function via the ``jac`` keyword argument. This is easy to include; we
+can simply create a function that computes the gradient using ``qml.grad``:
+
+>>> minimize(cost, params, method='BFGS', jac=lambda x: qml.grad(cost)(x)[0])
```
Comment (josh146) [inline doc/introduction/interfaces/numpy.rst:366]: The need to wrap the gradient function with `lambda` and index into element 0 will be unnecessary once #1081 is merged, and can simply be written `jac=qml.grad(cost)`.

**Comment by josh146** [comment]
[ch1202]

File: doc/introduction/interfaces.rst [hunk 3]
```diff
@@ -3,17 +3,99 @@
...
+However, for rapid prototyping on simulators, these methods typically out-perform forward-mode
+accumulators such as the parameter-shift rule and finite-differences. For more details, see the
+:doc:`quantum backpropagation <demos/tutorial_backprop>` demonstration.
+
+* ``"backprop"``: Use standard backpropagation.
+
+  This differentiation method is only allowed on simulator
+  devices that are classically end-to-end differentiable, for example
+  :class:`default.qubit <~.DefaultQubit>`. This method does *not* work on devices
```
Comment (trbromley) [inline doc/introduction/interfaces.rst:58]: Are we purposely simplifying here (i.e., hiding the fact we swap out `default.qubit` for its differentiable versions)? I think it makes sense... :thinking:

File: doc/introduction/interfaces.rst [hunk 4]
```diff
@@ -3,17 +3,99 @@
...
+    Typically, PennyLane will prioritize device-provided gradients, backpropagation, parameter-shift
+    rule, and finally finite-differences, in that order.
+
+Device gradients
+~~~~~~~~~~~~~~~~
+
+* ``"device"``: Queries the device directly for the gradient.
+  Only allowed on devices that provide their own gradient computation.
+
+Reverse-mode differentiation
```
Comment (trbromley) [inline doc/introduction/interfaces.rst:43]: What about "Simulation-based differentiation"? Although I could be persuaded for reverse mode, but I'm worried it could be confusing.

File: doc/introduction/interfaces.rst (hunk 1 above)
Comment (mariaschuld) [reply doc/introduction/interfaces.rst:118]: It's a note, it will be great. I use quickstart all the time as a reference.

File: doc/introduction/interfaces/numpy.rst (hunk 2 above)
Comment (mariaschuld) [reply doc/introduction/interfaces/numpy.rst:366]: Huge relief that change!

File: doc/introduction/interfaces.rst [hunk 5]
```diff
@@ -3,17 +3,99 @@
 
 .. _intro_interfaces:
 
-Interfaces and training
-=======================
+Gradients and training
+======================
```
Comment (mariaschuld) [inline doc/introduction/interfaces.rst:7]: Like that!

File: doc/introduction/interfaces.rst [hunk 6]
```diff
@@ -3,17 +3,99 @@
...
 
 PennyLane offers seamless integration between classical and quantum computations. Code up quantum
-circuits in PennyLane, and connect them easily to the top scientific computing and machine learning
-libraries.
+circuits in PennyLane, compute :doc:`quantum gradients <glossary/quantum_gradient>`, and
```
Comment (mariaschuld) [inline doc/introduction/interfaces.rst:10]:
```suggestion
circuits in PennyLane, compute :doc:`gradients of quantum circuits <glossary/quantum_gradient>`, and
```
just a suggestion...I realised that "quantum + word" is actually quite confusing. What is quantum about the gradient, does it contain a special kind of magic gradient? So trying to be explicit where possible.

File: doc/introduction/interfaces.rst [hunk 7]
```diff
@@ -3,17 +3,99 @@
...
+
+PennyLane currently provides the following differentiation methods for QNodes, divided into
+the following three categories.
+
+.. note::
+
+    If not specified, the default differentiation method is ``diff_method="best"``. PennyLane
+    will attempt to determine the *best* differentiation method given the device and interface.
+    Typically, PennyLane will prioritize device-provided gradients, backpropagation, parameter-shift
+    rule, and finally finite-differences, in that order.
```
Comment (mariaschuld) [inline doc/introduction/interfaces.rst:35]: I would move this note to the bottom of section, one is expecting the three items now.

File: doc/introduction/interfaces.rst [hunk 8]
```diff
@@ -3,17 +3,99 @@
...
+
+Device gradients
+~~~~~~~~~~~~~~~~
```
Comment (mariaschuld) [inline doc/introduction/interfaces.rst:38]: Would move this to last, it is the least useable case...

File: doc/introduction/interfaces.rst [hunk 9]
```diff
@@ -3,17 +3,99 @@
...
+
+* ``"device"``: Queries the device directly for the gradient.
+  Only allowed on devices that provide their own gradient computation.
+
+Reverse-mode differentiation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (mariaschuld) [inline doc/introduction/interfaces.rst:44]: What about "Gradients from simulators" or "simulator-only gradients" and below "Hardware-compatible gradients"?

File: doc/introduction/interfaces.rst [hunk 10]
```diff
@@ -3,17 +3,99 @@
...
+  the ``parameter-shift`` rule instead.
+
+* ``"adjoint"``: A form of backpropagation that takes advantage of the unitary or reversible
+  nature of quantum computation.
+
+  The `adjoint method <https://arxiv.org/abs/2009.02823>`__  reverses through the circuit after a
+  forward pass by iteratively applying the inverse (adjoint) gate. This method is similar to
+  ``"backprop"``, but trades off increased runtime with significantly lower memory usage.
+
+* ``"reversible"``: A form of backpropagation that takes advantage of the unitary or reversible
```
Comment (mariaschuld) [inline doc/introduction/interfaces.rst:69]: Reverse-mode and reversible could be confused..

File: doc/introduction/interfaces.rst [hunk 11]
```diff
@@ -3,17 +3,99 @@
...
+  nature of quantum computation.
+
+  This method is similar to the ``adjoint`` method, but has a lower time overhead and a similar
+  memory overhead.
```
Comment (mariaschuld) [inline doc/introduction/interfaces.rst:73]: One would wonder why we should use `adjoint` at all then?

File: doc/introduction/interfaces.rst [hunk 12]
```diff
@@ -3,17 +3,99 @@
...
+
+  Compared to the parameter-shift rule, the reversible method can be faster or slower, depending on
+  the density and location of parametrized gates in a circuit.
+
+Forward-mode differentiation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following methods support both quantum hardware and simulators. However, when using a simulator,
+you may notice that the computational time required to compute the gradients scales
```
Comment (mariaschuld) [inline doc/introduction/interfaces.rst:82]:
```suggestion
you may notice that the time required to compute the gradients scales
```

File: doc/introduction/interfaces.rst [hunk 13]
```diff
@@ -3,17 +3,99 @@
...
+
+
+Training and interfaces
+-----------------------
 
 The bridge between the quantum and classical worlds is provided in PennyLane via *interfaces*.
-Currently, there are four built-in interfaces: NumPy, PyTorch, JAX, and TensorFlow.
-These interfaces make each of these libraries quantum-aware, allowing quantum circuits to be
-treated just like any other operation.
+Currently, there are four built-in interfaces: NumPy, PyTorch, JAX, and TensorFlow. These interfaces
```
Comment (mariaschuld) [inline doc/introduction/interfaces.rst:96]: Links? :)

File: doc/introduction/interfaces.rst (hunk 1 above)
Comment (mariaschuld) [reply doc/introduction/interfaces.rst:118]: I'd move it to the bottom of the page though?

File: doc/introduction/interfaces/numpy.rst (hunk 2 above)
Comment (mariaschuld) [reply doc/introduction/interfaces/numpy.rst:366]: Nice addition!

**Review by mariaschuld — COMMENTED** [review]
Awesome Josh, this is very useful indeed.

I had a few comments, too many I guess to approve yet. Altogether I think the diff method section could either link to or contain more information on the actual implementation - simply because the technical terms like reverse-mode and adjoint are not yet common. 

Two things would help: a) links to papers or b) equations to show the principle? Or entries in the glossary? 

I would myself actually love to know this, I never understood the difference btw adjoint and reverse very well.

File: doc/introduction/interfaces.rst (hunk 11 above)
Comment (josh146) [reply doc/introduction/interfaces.rst:73]: Oh good catch, this is wrong. Should be 'has a _larger_ time overhead' 🤦

File: doc/introduction/interfaces.rst [hunk 14]
```diff
@@ -3,17 +3,99 @@
...
+  nature of quantum computation.
+
+  The `adjoint method <https://arxiv.org/abs/2009.02823>`__  reverses through the circuit after a
+  forward pass by iteratively applying the inverse (adjoint) gate. This method is similar to
+  ``"backprop"``, but trades off increased runtime with significantly lower memory usage.
+
+* ``"reversible"``: A form of backpropagation that takes advantage of the unitary or reversible
+  nature of quantum computation.
+
+  This method is similar to the ``adjoint`` method, but has a lower time overhead and a similar
```
Comment (josh146) [inline doc/introduction/interfaces.rst:72]:
```suggestion
  This method is similar to the ``adjoint`` method, but has a slightly larger time overhead and a similar
```

File: doc/introduction/interfaces.rst (hunk 9 above)
Comment (josh146) [reply doc/introduction/interfaces.rst:44]: Oh, yes good idea. I think this is nicer and more relevant.

**Comment by josh146** [comment]
> I had a few comments, too many I guess to approve yet. Altogether I think the diff method section could either link to or contain more information on the actual implementation - simply because the technical terms like reverse-mode and adjoint are not yet common.
>
> Two things would help: a) links to papers or b) equations to show the principle? Or entries in the glossary?

I agree that we should include more details about the working on these methods. I spent a long time thinking about this while writing the PR, I'm just not sure it makes sense to put this in the quickstart/PL API docs?

I think this makes more sense as glossary/key concept entries, like the [parameter-shift entry](https://pennylane.ai/qml/glossary/parameter_shift.html). We can then link to the glossary directly from the quickstart.

Another advantage to putting it in the glossary is we can update it at any time, independent of a PL release.

File: doc/introduction/interfaces.rst (hunk 11 above)
Comment (trbromley) [inline doc/introduction/interfaces.rst]: It seems like these have been swapped :thinking: Isn't it the adjoint method that is similar to backprop, but with lower time overhead and similar memory overhead?

File: doc/introduction/interfaces.rst [hunk 15]
```diff
@@ -3,17 +3,99 @@
...
+
+  Compared to the parameter-shift rule, the reversible method can be faster or slower, depending on
+  the density and location of parametrized gates in a circuit.
+
+Forward-mode differentiation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following methods support both quantum hardware and simulators. However, when using a simulator,
+you may notice that the time required to compute the gradients scales
+with the number of trainable circuit parameters.
```
Comment (trbromley) [inline doc/introduction/interfaces.rst:83]: It's also true that the time scales with number of params for methods like adjoint. Could we just say that it may be slower on a simulator to use forward mode?

File: doc/introduction/interfaces.rst (hunk 1 above)
Comment (trbromley) [reply doc/introduction/interfaces.rst:118]: I like it too!

**Comment by mariaschuld** [comment]
> > I had a few comments, too many I guess to approve yet. Altogether I think the diff method section could either link to or contain more information on the actual implementation - simply because the technical terms like reverse-mode and adjoint are not yet common.
> > Two things would help: a) links to papers or b) equations to show the principle? Or entries in the glossary?
> 
> I agree that we should include more details about the working on these methods. I spent a long time thinking about this while writing the PR, I'm just not sure it makes sense to put this in the quickstart/PL API docs?
> 
> I think this makes more sense as glossary/key concept entries, like the [parameter-shift entry](https://pennylane.ai/qml/glossary/parameter_shift.html). We can then link to the glossary directly from the quickstart.
> 
> Another advantage to putting it in the glossary is we can update it at any time, independent of a PL release.

Makes sense!

**Review by trbromley — COMMENTED** [review]
Thanks @josh146 this looks like very useful content in the docs! I left a few comments but will be happy to approve after.

File: doc/introduction/interfaces.rst [hunk 16]
```diff
@@ -3,17 +3,99 @@
 
 .. _intro_interfaces:
 
-Interfaces and training
-=======================
+Gradients and training
```
[the remark this answers is not in the document]
Comment (thisac) [reply doc/introduction/interfaces.rst:6]: I would lean towards "Gradients and interfaces" as well here.

File: doc/introduction/interfaces.rst [hunk 17]
```diff
@@ -3,17 +3,99 @@
...
+
+.. code-block:: python
+
+    @qml.qnode(dev, diff_method="parameter-shift")
+    def circuit(x):
+        qml.RX(x, wires=0)
+        return qml.probs(wires=0)
+
+PennyLane currently provides the following differentiation methods for QNodes, divided into
+the following three categories.
```
Comment (thisac) [inline doc/introduction/interfaces.rst:28]: Maybe reword this so that "following" is only used once. Something like: "PennyLane provides a few different differentiation methods for QNodes, divided into the following categories."

File: doc/introduction/interfaces.rst (hunk 3 above)
Comment (thisac) [reply doc/introduction/interfaces.rst:58]: I think that makes sense as well, since this is only a simplified overview.

File: doc/introduction/interfaces.rst (hunk 16 above)
[the remark this answers is not in the document]
Comment (josh146) [reply doc/introduction/interfaces.rst:6]:
```suggestion
Gradients and interfaces
```

Although be good to get @mariaschuld's thoughts before switching it!

File: doc/introduction/interfaces.rst [hunk 18]
```diff
@@ -3,17 +3,99 @@
...
+
+.. note::
+
+    If not specified, the default differentiation method is ``diff_method="best"``. PennyLane
+    will attempt to determine the *best* differentiation method given the device and interface.
+    Typically, PennyLane will prioritize device-provided gradients, backpropagation, parameter-shift
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/introduction/interfaces.rst:34]: I just wanted to provide a buffer in case the logic ever does something more complicated 😆

File: doc/introduction/interfaces.rst (hunk 3 above)
Comment (josh146) [reply doc/introduction/interfaces.rst:58]: Yep, I'm purposely simplifying :) I think this is a case where providing the full details will just be more confusing.

I'm thinking we should move towards a case where `default.qubit.tf` etc. are no longer mentioned; instead, we simply refer to `default.qubit` supporting tf, autograd, jax.

File: doc/introduction/interfaces.rst (hunk 15 above)
Comment (josh146) [reply doc/introduction/interfaces.rst:83]: Wait really? I didn't know that. I thought the adjoint method was a reverse accumulator, in that it scales with the dimension of the output, not the input of the function being differentiated?

Comment (josh146) [reply doc/introduction/interfaces.rst:83]: Oh, there is a slightly subtlety here I think?

for a function R^m -> R^n, 

* forward accumulators (finite diff, param shift) require `m` quantum evaluations.

* Reverse accumulators instead require 'n' quantum evaluations (this can be seen somewhat from the line

  ```python
  lambdas = [self._apply_operation(lambda_, op) for lambda_ in lambdas]
  ```

  in the adjoint logic, where `len(lambda) ~ tape.output_dim`).

But the reverse accumulators will still somewhat scale with the number of parameters insomuch as more parameters = deeper circuits? And in the adjoint case, more generators to compute?

File: doc/introduction/interfaces.rst (hunk 16 above)
[the remark this answers is not in the document]
Comment (mariaschuld) [reply doc/introduction/interfaces.rst:6]: I really like training, because, the interfaces _do_ the training...it describes the page by what it is used for, not by a technical PL lingo term.

[the remark this answers is not in the document]
Comment (josh146) [reply doc/introduction/interfaces.rst:6]: I agree. 'Gradients and Training' will mean something to new users looking to learn PL, and as a side effect they'll learn what 'interface' means.

Whereas 'Gradients and Interfaces' is a bit more esoteric sounding.

File: doc/introduction/interfaces.rst (hunk 8 above)
Comment (josh146) [reply doc/introduction/interfaces.rst:38]: done!

File: doc/introduction/interfaces.rst (hunk 4 above)
Comment (josh146) [reply doc/introduction/interfaces.rst:43]: Nice, changed this!

File: doc/introduction/interfaces.rst (hunk 9 above)
Comment (josh146) [reply doc/introduction/interfaces.rst:44]: Went with "Simulation-based differentiation" as per Tom's suggestion!

File: doc/introduction/interfaces.rst (hunk 15 above)
Comment (josh146) [reply doc/introduction/interfaces.rst:83]: Changed this sentence to 'scales quadratically' to be more specific.

File: doc/introduction/interfaces.rst (hunk 1 above)
Comment (josh146) [reply doc/introduction/interfaces.rst:118]: Have moved it to the bottom of the page!

File: doc/introduction/interfaces.rst (hunk 13 above)
Comment (josh146) [reply doc/introduction/interfaces.rst:96]: done!

**Comment by josh146** [comment]
Thanks everyone, have incorporated suggestions and it's already looking a lot better!

File: doc/introduction/interfaces.rst [hunk 19]
```diff
@@ -3,17 +3,106 @@
...
+
+.. code-block:: python
+
+    @qml.qnode(dev, diff_method="parameter-shift")
+    def circuit(x):
+        qml.RX(x, wires=0)
+        return qml.probs(wires=0)
+
+PennyLane currently provides the following differentiation methods for QNodes, divided into
+the following categories.
```
Comment (mariaschuld) [inline doc/introduction/interfaces.rst:28]:
```suggestion
PennyLane currently provides the following differentiation methods for QNodes:
```
(repetition of word "following")

File: doc/introduction/interfaces.rst [hunk 20]
```diff
@@ -3,17 +3,106 @@
...
+
+Simulation-based differentiation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following methods use `reverse accumulation
+<https://en.wikipedia.org/wiki/Automatic_differentiation#Reverse_accumulation>`__ to compute
+gradients; a well-known example of this approach is backpropagation. These methods only support
+state-vector simulators, they are **not** hardware compatible; they are only supported on
+*statevector* simulator devices such as :class:`default.qubit <~.DefaultQubit>`.
```
Comment (mariaschuld) [inline doc/introduction/interfaces.rst:37]:
```suggestion
gradients; a well-known example of this approach is backpropagation. These methods are **not** hardware compatible; they are only supported on
*statevector* simulator devices such as :class:`default.qubit <~.DefaultQubit>`.
```

File: doc/introduction/interfaces.rst (hunk 11 above)
Comment (mariaschuld) [reply doc/introduction/interfaces.rst:62]: Wait but if it is similar on memory and slower, why would one use it?

**Review by mariaschuld — APPROVED** [review]
Only two more changes of language...nice!

File: doc/introduction/interfaces.rst [hunk 21]
```diff
@@ -3,17 +3,99 @@
...
+  the ``parameter-shift`` rule instead.
+
+* ``"adjoint"``: A form of backpropagation that takes advantage of the unitary or reversible
+  nature of quantum computation.
+
+  The `adjoint method <https://arxiv.org/abs/2009.02823>`__  reverses through the circuit after a
+  forward pass by iteratively applying the inverse (adjoint) gate. This method is similar to
+  ``"backprop"``, but trades off increased runtime with significantly lower memory usage.
+
+* ``"reversible"``: A form of backpropagation that takes advantage of the unitary or reversible
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/introduction/interfaces.rst:58]: Not yet :( We should add a page to the glossary section, that goes through the mathematics and has examples.

File: doc/introduction/interfaces.rst [hunk 22]
```diff
@@ -3,17 +3,104 @@
...
+  nature of quantum computation.
+
+  This method is similar to the ``adjoint`` method, but has a slightly larger time overhead and a similar
+  memory overhead.
+
+  Compared to the parameter-shift rule, the reversible method can be faster or slower, depending on
+  the density and location of parametrized gates in a circuit.
```
Comment (thisac) [inline doc/introduction/interfaces.rst:63]: (minor) I personally think it looks better combining these two paragraphs into one.

```suggestion
  This method is similar to the ``adjoint`` method, but has a slightly larger time overhead and a similar
  memory overhead. Compared to the parameter-shift rule, the reversible method can be faster or slower,
  depending on the density and location of parametrized gates in a circuit.
```

File: doc/introduction/interfaces.rst [hunk 23]
```diff
@@ -3,17 +3,104 @@
...
+The following methods use `reverse accumulation
+<https://en.wikipedia.org/wiki/Automatic_differentiation#Reverse_accumulation>`__ to compute
+gradients; a well-known example of this approach is backpropagation. These methods are **not** hardware compatible; they are only supported on
+*statevector* simulator devices such as :class:`default.qubit <~.DefaultQubit>`.
+
+However, for rapid prototyping on simulators, these methods typically out-perform forward-mode
+accumulators such as the parameter-shift rule and finite-differences. For more details, see the
+:doc:`quantum backpropagation <demos/tutorial_backprop>` demonstration.
+
+* ``"backprop"``: Use standard backpropagation.
```
Comment (thisac) [inline doc/introduction/interfaces.rst:41]: (minor) "backprop" uses imperative form in it's first line description (same with "parameter shift" and "device"), while neither "adjoint" nor "reversible" do. It might look slightly better to always use imperative here?

File: doc/introduction/interfaces.rst [hunk 24]
```diff
@@ -3,17 +3,104 @@
...
+accumulation <https://en.wikipedia.org/wiki/Automatic_differentiation#Forward_accumulation>`__
+However, when using a simulator, you may notice that the time required to compute the gradients
+:doc:`scales quadratically <demos/tutorial_backprop>` with the number of trainable circuit
+parameters.
+
+* ``"parameter-shift"``: Use the analytic :doc:`parameter-shift rule
+  <glossary/parameter_shift>` for all supported quantum operation arguments, with
+  finite-difference as a fallback.
+
+* ``"finite-diff"``: Uses numerical finite-differences for all quantum operation arguments.
```
Comment (thisac) [inline doc/introduction/interfaces.rst:78]:
```suggestion
* ``"finite-diff"``: Use numerical finite-differences for all quantum operation arguments.
```

File: doc/introduction/interfaces.rst [hunk 25]
```diff
@@ -3,17 +3,104 @@
...
+  :class:`default.qubit <~.DefaultQubit>`. This method does *not* work on devices
+  that estimate measurement statistics using a finite number of shots; please use
+  the ``parameter-shift`` rule instead.
+
+* ``"adjoint"``: A form of backpropagation that takes advantage of the unitary or reversible
+  nature of quantum computation.
+
+  The `adjoint method <https://arxiv.org/abs/2009.02823>`__  reverses through the circuit after a
+  forward pass by iteratively applying the inverse (adjoint) gate. This method is similar to
+  ``"backprop"``, but trades off increased runtime with significantly lower memory usage.
```
Comment (trbromley) [inline doc/introduction/interfaces.rst:54]: Is this true? i.e., "trades off increased runtime with significantly lower memory usage"?
We saw on benchmarks that adjoint was faster, and I'm not sure if there's a complexity argument to say backprop is faster?

Note that this sentence seems closer to the one we have for reversible:
"The reversible approach is similar to backpropagation, but trades off extra computation for enhanced memory efficiency."

This is what we say for adjoint:
"This method is similar to the reversible method, but has a lower time overhead and a similar memory overhead."

Overall, I need to think about it, but I'm not sure if there is an occasion where reversible is preferable.

File: doc/introduction/interfaces.rst [hunk 26]
```diff
@@ -3,17 +3,104 @@
...
+
+* ``"reversible"``: A form of backpropagation that takes advantage of the unitary or reversible
```
Comment (trbromley) [inline doc/introduction/interfaces.rst:56]:
```suggestion
* ``"reversible"``: A form of differentiation that takes advantage of the unitary or reversible
```
Is it a form of backprop? (It might be, maybe I'm just confused)

File: doc/introduction/interfaces.rst [hunk 27]
```diff
@@ -3,17 +3,104 @@
...
+  memory overhead.
+
+  Compared to the parameter-shift rule, the reversible method can be faster or slower, depending on
+  the density and location of parametrized gates in a circuit.
+
+Hardware-compatible differentiation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following methods support both quantum hardware and simulators, and are examples of `forward
+accumulation <https://en.wikipedia.org/wiki/Automatic_differentiation#Forward_accumulation>`__
```
Comment (trbromley) [inline doc/introduction/interfaces.rst:69]:
```suggestion
accumulation <https://en.wikipedia.org/wiki/Automatic_differentiation#Forward_accumulation>`__.
```

File: doc/introduction/interfaces.rst [hunk 28]
```diff
@@ -3,17 +3,104 @@
...
+* ``"device"``: Queries the device directly for the gradient.
+  Only allowed on devices that provide their own gradient computation.
+
+
+.. note::
+
+    If not specified, the default differentiation method is ``diff_method="best"``. PennyLane
+    will attempt to determine the *best* differentiation method given the device and interface.
+    Typically, PennyLane will prioritize device-provided gradients, backpropagation, parameter-shift
+    rule, and finally finite-differences, in that order.
```
Comment (trbromley) [inline doc/introduction/interfaces.rst:93]: We can revisit this when adjoint becomes the default.

File: doc/introduction/interfaces.rst [hunk 29]
```diff
@@ -3,17 +3,104 @@
...
+-----------------------
 
 The bridge between the quantum and classical worlds is provided in PennyLane via *interfaces*.
-Currently, there are four built-in interfaces: NumPy, PyTorch, JAX, and TensorFlow.
-These interfaces make each of these libraries quantum-aware, allowing quantum circuits to be
-treated just like any other operation.
+Currently, there are four built-in interfaces: :doc:`NumPy <interfaces/numpy>`, :doc:`PyTorch
+<interfaces/torch>`, :doc:`JAX <interfaces/jax>`, and :doc:`TensorFlow <interfaces/tf>`. These
+interfaces make each of these libraries quantum-aware, allowing quantum circuits to be treated just
+like any other operation.
```
Comment (thisac) [inline doc/introduction/interfaces.rst:103]: (minor) Below, it says: `(NumPy arrays, Torch Tensors, or TensorFlow Tensors)`. Might make more sense to use lowercase for `Tensors`, since it refers to the type/word tensor, rather than the `Tensor` object, right? 🤔

File: doc/introduction/interfaces.rst [hunk 30]
```diff
@@ -96,6 +183,15 @@ converting QNodes into both Keras and ``torch.nn`` layers:
     pennylane.qnn.KerasLayer
     pennylane.qnn.TorchLayer
 
+
+.. note::
+
+    QNodes with an interface will always incur a small overhead on evaluation. If you do not
+    need to compute quantum gradients of a QNode, specifying ``interface=None`` will remove
+    this overhead and result in a slightly faster evaluation. However, gradients will no
+    longer be available.
```
Comment (thisac) [inline doc/introduction/interfaces.rst:192]: 💯

File: doc/introduction/interfaces/numpy.rst (hunk 2 above)
Comment (thisac) [inline doc/introduction/interfaces/numpy.rst:366]: @josh146 Don't forget to change this here if you merge #1081 first!
```suggestion
>>> minimize(cost, params, method='BFGS', jac=qml.grad(cost))
```

**Review by trbromley — APPROVED** [review]
Thanks @josh146, nice addition to the docs!

**Review by thisac — APPROVED** [review]
Looks great @josh146! 💯 Just a few very minor comments.

File: doc/introduction/interfaces.rst (hunk 26 above)
Comment (josh146) [reply doc/introduction/interfaces.rst:56]: Ah... I believe that `reversible` is a form of backpropagation, in that backpropagation refers to reverse-mode gradient accumulation of a scalar-valued function?

File: doc/introduction/interfaces.rst [hunk 31]
```diff
@@ -3,17 +3,104 @@
...
+
+* ``"backprop"``: Use standard backpropagation.
+
+  This differentiation method is only allowed on simulator
+  devices that are classically end-to-end differentiable, for example
+  :class:`default.qubit <~.DefaultQubit>`. This method does *not* work on devices
+  that estimate measurement statistics using a finite number of shots; please use
+  the ``parameter-shift`` rule instead.
+
+* ``"adjoint"``: A form of backpropagation that takes advantage of the unitary or reversible
```
Comment (josh146) [inline doc/introduction/interfaces.rst:49]:
```suggestion
* ``"adjoint"``: Use a form of backpropagation that takes advantage of the unitary or reversible
```

File: doc/introduction/interfaces.rst (hunk 26 above)
Comment (josh146) [inline doc/introduction/interfaces.rst:56]:
```suggestion
* ``"reversible"``: Use a form of backpropagation that takes advantage of the unitary or reversible
```

File: doc/introduction/interfaces/numpy.rst (hunk 2 above)
Comment (josh146) [reply doc/introduction/interfaces/numpy.rst:366]: Good catch!

File: doc/introduction/interfaces.rst (hunk 25 above)
Comment (josh146) [reply doc/introduction/interfaces.rst:54]: @trbromley this is a good point. Do you have any suggestions we should change it to?

Comment (trbromley) [reply doc/introduction/interfaces.rst:54]: :thinking: Could we steer clear of making comparisons between the methods? I feel like probably only one option should be on offer eventually, with the others potentially available but not advertised. We could e.g., do a tutorial comparing?

But in the meantime, yes maybe we don't compare, or:
"This method is similar to ``"backprop"``, but has significantly lower memory usage and a similar runtime."

Actually, what is the runtime complexity of backprop?

Comment (josh146) [reply doc/introduction/interfaces.rst:54]:
```suggestion
  ``"backprop"``, but has significantly lower memory usage and a similar runtime.
```

> Actually, what is the runtime complexity of backprop?

I am not 100% sure. I know it tends to have a constant overhead compared to the forward pass, with an additional memory requirement during the forward pass due to storing intermediate computations

Comment (trbromley) [reply doc/introduction/interfaces.rst:54]: Great, yes if it has a constant overhead relative to the forward pass then it is similar complexity to the adjoint method, which is effectively 2 forward passes (one forward, one in reverse) - although I would need to think more carefully before being totally confident about this statement :laughing:.

Comment (josh146) [reply doc/introduction/interfaces.rst:54]: This definitely makes sense then --- with the adjoint method, we are not storing the intermediate computations (thus, less memory), _but_ we need to 'rewind' a few operations (thus, similar/slightly longer runtime)
